### PR TITLE
Support Multi Sub Targets

### LINKED ISSUE #423: Rust WASM toolchain not installed, please install it
I'm working on substrate. It'll also compile a WASM part at the same time.

```rust
Caused by:
  process didn't exit successfully: `/target/release/build/crab-runtime-b00ee5d1b3acfcb8/build-script-build` (exit code: 1)
--- stderr
   Compiling wasm-build-runner-impl v1.0.0 (/target/x86_64-unknown-linux-gnu/release/wbuild-runner/crab-runtime7359505889082344466)
    Finished release [optimized] target(s) in 13.61s
     Running `/linux-runner x86_64 /target/x86_64-unknown-linux-gnu/release/wbuild-runner/crab-runtime7359505889082344466/target/x86_64-unknown-linux-gnu/release/wasm-build-runner-impl`
Rust WASM toolchain not installed, please install it!

warning: build failed, waiting for other jobs to finish...
error: build failed
```

### PR DESCRIPTION
**Pull request by AurevoirXavier**
close #423

### CODE REVIEW

**Comment by AurevoirXavier** [comment]
This time I got

```rust
The following warnings were emitted during compilation:

warning: couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
warning: set the LLVM_CONFIG_PATH environment variable to the full path to a valid `llvm-config` executable (including the executable itself)

error: failed to run custom build command for `librocksdb-sys v6.7.4`

Caused by:
  process didn't exit successfully: `/target/release/build/librocksdb-sys-aaf9101b7f66f631/build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=build.rs
  cargo:warning=couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
  cargo:warning=set the LLVM_CONFIG_PATH environment variable to the full path to a valid `llvm-config` executable (including the executable itself)

  --- stderr
  thread 'main' panicked at 'Unable to find libclang: "couldn\'t find any valid shared libraries matching: [\'libclang.so\', \'libclang-*.so\', \'libclang.so.*\', \'libclang-*.so.*\'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"', /cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.53.3/src/lib.rs:1956:31
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by AurevoirXavier** [comment]
```cross build --release --target x86_64-unknown-linux-gnu --sub-targets wasm32-unknown-unknown``` 

Perfect.

**Comment by AurevoirXavier** [comment]
@reitermarkus Please review.

**Comment by reitermarkus** [comment]
Is this still relevant? It looks like the only thing this does is install a target using `rustup`. Not sure this warrants adding a flag.

**Comment by AurevoirXavier** [comment]
> Is this still relevant? It looks like the only thing this does is install a target using `rustup`. Not sure this warrants adding a flag.

Some projects need multiple targets. I'm curious, without this flag how can we do that?

**Comment by AurevoirXavier** [comment]
> > without this flag how can we do that?
> 
> Compile every target separately?
> 
> ```
> cross build --release --target x86_64-unknown-linux-gnu
> cross build --release --target wasm32-unknown-unknown
> ```

Nope. The wasm32 target was compiled from a `build.rs`.

**Comment by Emilgardis** [comment]
Could you show us the full invokation that errors without this CLI flag and a way for us to reproduce it @AurevoirXavier

**Comment by AurevoirXavier** [comment]
> Could you show us the full invokation that errors without this CLI flag and a way for us to reproduce it @AurevoirXavier

```
git clone https://github.com/paritytech/substrate.git && cd substrate && cross build --target x86_64-unknown-linux-gnu
```

**Comment by Alexhuszagh** [comment]
> > Is this still relevant? It looks like the only thing this does is install a target using `rustup`. Not sure this warrants adding a flag.
> 
> Some projects need multiple targets. I'm curious, without this flag how can we do that?

I don't think this works either, since we'd have to compile every target separately anyway since we don't assume the cross-compiler toolchain is installed on the host, and it definitely isn't installed on the image. I think the only way is to build for multiple targets. If you need multiple targets built at once, just write a wrapper script?

**Comment by Alexhuszagh** [comment]
I'm willing to reopen this if there's a clear purpose this would solve and a way of implementing it other than just fetching all the images separately, and then running cargo for that target in each separate image. As it stands, I can't see this being any better than:

```bash
targets=(x86_64-unknown-linux-gnu wasm32-unknown-unknown)
for target in "${targets[@]}"; do
     cross build --target "${target}"
done
```

Which I really don't think warrants this.

**Comment by AurevoirXavier** [comment]
As I said, the other target was built from the `build.rs`. We can not build it directly/separately.
For the toolchain, I can use `rust-toolchain.toml` to manage.

Did you try the project I link?
Looks like not, I think you guys are so arrogant.
Never mind, I'll keep using my fork version.

**Comment by Alexhuszagh** [comment]
> Did you try the project I link? Looks like not, I think you guys are so arrogant. Never mind, I'll keep using my fork version.

Please be polite. I also mentioned I'm willing to re-open this if there's a clear rationale and method to implement this, because right now I can't see it.

> I'll say sorry if your proposal can work on that project. But I had tried that 2 years ago.

I've tried the project. Well, it has bindgen issue which I'm testing now but I don't foresee any issues. Also, your code for the subtargets just installed the Rust toolchains, but that doesn't mean anything else which is required to use the toolchain is installed, which we cannot do. We cannot create every combination of every toolchain into an image, and due to some path conflicts, we can't create an image with every toolchain and it would be gigantic as well.

This is the primary logic of the code, which installs the toolchain for Rustup, but this does not include the C compiler, linker, C++ compiler, or C standard library: in short, it lacks all the stuff required to compile Rust crates for another target.

Here's the major change:

https://github.com/cross-rs/cross/pull/448/commits/4833941b008ab25ecea9b9d0fd00c0b817b5deb4#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR287-R298

Or as code:

```rust
if !uses_xargo {
    if let Some(targets) = args.sub_targets {
        for target in targets {
            if !available_targets.is_installed(&target)
                && available_targets.contains(&target)
            {
                rustup::install(&target, &toolchain, verbose)?;
            }
        }
    }
}
```

For example, if I add:

```bash
$ rustup target add mips64-unknown-linux-gnuabi64
$ cargo build --target  mips64-unknown-linux-gnuabi64
   Compiling cc v1.0.73
   Compiling hellopp v0.1.0 (C:\Users\ahusz\git\rust-cpp-hello-word)
error: failed to run custom build command for `hellopp v0.1.0 (C:\Users\ahusz\git\rust-cpp-hello-word)`

Caused by:
  process didn't exit successfully: `C:\Users\ahusz\git\rust-cpp-hello-word\target\debug\build\hellopp-e3257b093cd0588e\build-script-build` (exit code: 1)
  --- stdout
  TARGET = Some("mips64-unknown-linux-gnuabi64")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-pc-windows-msvc")
  CXX_mips64-unknown-linux-gnuabi64 = None
  CXX_mips64_unknown_linux_gnuabi64 = None
  TARGET_CXX = None
  CXX = None
  CROSS_COMPILE = None
  CXXFLAGS_mips64-unknown-linux-gnuabi64 = None
  CXXFLAGS_mips64_unknown_linux_gnuabi64 = None
  TARGET_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = None
  running: "mips64-linux-gnuabi64-g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-Wall" "-Wextra" "-o" "C:\\Users\\ahusz\\git\\rust-cpp-hello-word\\target\\mips64-unknown-linux-gnuabi64\\debug\\build\\hellopp-3ad811bd01d5673d\\out\\hellopp.o" "-c" "hellopp.cc"

  --- stderr


  error occurred: Failed to find tool. Is `mips64-linux-gnuabi64-g++` installed? (see https://github.com/alexcrichton/cc-rs#compile-time-requirements for help)
```

This is in a crate with a C++ dependency, and we cannot find the cross-compiler for C++, as expected. But let's try a simpler crate:

```bash
$ cargo init --bin hello
$ cd hello
$ rustup target add mips64-unknown-linux-gnuabi64
$ cargo build --target  mips64-unknown-linux-gnuabi64
   Compiling hello v0.1.0 (C:\Users\ahusz\git\hello)
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.1p4bv2i9hufp5zx.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.26t5c79bq1h6ur8d.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.2tp9rbleejrfe2x4.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.3tqyicmrj57llvj9.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.56c6bx8f0r48ao4h.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.57caedyxp1gpmk7o.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.6vqgsaw4sofan4i.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.nasy1m8qmgm3dc3.rcgu.o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9.4r5mp1d9bu1s3v3z.rcgu.o" "-Wl,--as-needed" "-L" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps" "-L" "C:\\Users\\ahusz\\git\\hello\\target\\debug\\deps" "-L" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib" "-Wl,--start-group" "-Wl,-Bstatic" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libstd-d0aef713850b46dc.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libpanic_unwind-e189549181b9a451.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libobject-7b802252dd7985f2.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libmemchr-cbc185d82bb8700a.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libaddr2line-7ddca690744a49ee.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libgimli-27639e86e91301e4.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\librustc_demangle-b22402584f32474b.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libstd_detect-5be9fa60f68a342a.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libhashbrown-c920e7723f3fa345.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libminiz_oxide-294ec455fab9a908.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libadler-945ec988bbfbe2c7.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\librustc_std_workspace_alloc-c602b03fbf63f88c.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libunwind-36536e1f9e284d39.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libcfg_if-8b8fdf8c6ef37fc8.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\liblibc-bc4f8e5bae134bb2.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\liballoc-30f332dc7535e228.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\librustc_std_workspace_core-4164a5bfc685074d.rlib" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libcore-8e1ddd0a9840fa11.rlib" "-Wl,--end-group" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib\\libcompiler_builtins-1ca08a292ce17a72.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "C:\\Users\\ahusz\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\mips64-unknown-linux-gnuabi64\\lib" "-o" "C:\\Users\\ahusz\\git\\hello\\target\\mips64-unknown-linux-gnuabi64\\debug\\deps\\hello-c84295074f2d50d9" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs"
  = note: C:/SysGCC/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: unrecognized option '--eh-frame-hdr'
          C:/SysGCC/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/9.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: use the --help option for usage information
          collect2.exe: error: ld returned 1 exit status


error: could not compile `hello` due to previous error
```

Here we can't find the linker, which is always going to be an issue, since the linker for the toolchain isn't installed.

In short, this solution **cannot** work as you've proposed it, without providing your own toolchain in the image or mounted into the image, in which case you can always add code to ensure cargo has the toolchain installed.

**Comment by AurevoirXavier** [comment]
I'm polite, and I'm patient.
I opened this two years ago.
And I gave you an example.

But I only get 'why not build every target separately' again and again.
If you try that, you will know that not going to work.

After that, if you point out that my solution is not general enough I'll totally accept it.

And thanks for your time.

**Comment by Alexhuszagh** [comment]
I was able to get it to compile (I think, it takes hours, I've fixed all the aforementioned issues though that were initially causing compilation errors) with the following:

1. An extension Dockerfile for bindgen (see the wiki).
2. `rustup target add wasm32-unknown-unknown`
3. `cross +nightly build --target x86_64-unknown-linux-musl`

I'm using `x86_64-unknown-linux-musl` because the `x86_64-unknown-linux-gnu` is CentOS which has an older Clang version that makes it so Clang3.9+ must be installed from source.

The `Dockerfile` is:

```Dockerfile
FROM ghcr.io/cross-rs/x86_64-unknown-linux-musl:main
RUN apt-get update && apt-get install --assume-yes --no-install-recommends libclang-dev clang
```

This is built with:

```bash
podman build --file Dockerfile --tag x86_64-unknown-linux-musl:local .
```

And then `Cross.toml` uses the image for the target:

```toml
[target.x86_64-unknown-linux-musl]
image = "x86_64-unknown-linux-musl:local"
```

The `wasm32-unknown-unknown` target can be compiled on the host: no extra toolchain required without external dependencies (just test `cargo init --bin hello && cd hello && cargo build --target wasm32-unknown-unknown`). So this example is the rare (possibly only) case in which this PR works?

Also, we've lacked maintainers for a long period of time, so we've been running on a major backlog. We've recently finally been able to get to older issues.

**Comment by AurevoirXavier** [comment]
Sorry, I still did not get the point.

`cargo build --target wasm32-unknown-unknown` are not going to work for substrate-based projects. It uses a build.rs, and constructs so many flags to build the wasm file. Then integrate that wasm file into the binary.

It looks like `executable(client + runtime.wasm) file` not `binary file + wasm file`.

**Comment by Alexhuszagh** [comment]
> `cargo build --target wasm32-unknown-unknown` are not going to work for substrate-based projects. It uses a build.rs, and constructs so many flags to build the wasm file. Then integrate that wasm file into the binary.

No that's the point of it being able to use the toolchain: that's just a sample crate to show we have basic WASM support (for Rust-only crates). It's showing `rustup target add ...` is sufficient for that target, but likely no others. And if there are C or C++ dependencies for code that needs to be compiled to WASM, then this would fail too. Obviously, the code would be mostly compiled to x86_64 with the above target.

This is the relevant code (which is what fails befre) just checking that the target has been added to Rustup, so it can compile Rust-only code to WASM:

https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/utils/wasm-builder/src/prerequisites.rs#L47-L70

**Comment by Alexhuszagh** [comment]
Does that make sense? Say we tried to embed a `x86_64-unknown-linux-gnu` binary in an `aarch64-unknown-linux-gnu` one (doesn't make much logical sense, but it's another example), we would need to run inside an image with the toolchains (C/C++ compiler, linker/archiver, and potentially C and C++ standard library), and we only have pre-built toolchains for one or the other. In short, without a combination image, it would fail, which is a nightmare maintenance wise.

This is a rare exception, because `wasm32-unknown-unknown` doesn't actually need any other tools installed for compiling Rust-only code, so you can compile for x86_64 and wasm within the x86_64 image without issue.

**Comment by AurevoirXavier** [comment]
Yes.

I figure out another way to archive this.
1. Build the https://github.com/paritytech/substrate/blob/master/bin/node/runtime/build.rs.
1. Put the wasm file to https://github.com/paritytech/substrate/blob/master/bin/node-template/runtime/src/lib.rs#L7.
1. Pass https://paritytech.github.io/substrate/master/src/substrate_wasm_builder/lib.rs.html#116 to the compiler.
